Limit braille cursor blink rate

feerrenrut opened the pull request:
Fix for #6470, limit the minimum blink rate to 200ms

This stops the blink rate from being set below 200ms. Configurations that are already set below 200ms will be modified, and set to 200ms.

There is some active discussion about being able to disable the braille cursor blinking, this was previously possible by setting the blink rate to 0.

feerrenrut commented:
@jcsteh Could you take a look at this? Note the questions around disabling the cursor blinking.

feerrenrut commented:
@jcsteh I have updated this so that there is an option to disable the cursor blinking. Since I dont have a braille display I cant really test that this works, other than the config saving / GUI side of things.

jcsteh reviewed (commented):
I'm fine with the GUI changes here.

Even though specifying 0 to disable blinking was never documented, it was definitely supported and I vaguely recall it was intentional. If so, there may be users using this. I'm wondering if we should keep the single setting in the config with a minimum of 0 and just do the separate boolean/integer with min 200 in the GUI. That way, we preserve backwards compat. That does mean the config spec doesn't enforce the minimum, though, which is a bit ugly.

There actually is a way to get the minimum value from the config spec, but it's ugly. Still, maybe we can write a function in config to hide the ugliness:

```
int(config.conf.validator._parse_with_caching(config.conf.spec["braille"]["cursorBlinkRate"])[2]["min"])
```

feerrenrut commented:
Yep, good point. With the current approach we will lose the users "dont blink" setting and replace it with blink at 200ms. I'll look to see if there is a way that we can return the actual value (perhaps along with the exception) when the value is below the minimum. This way we get to keep the matching config spec, and upgrade the settings. In order to provide a cleaner approach perhaps its worth talking about an "upgrade" step for the config. Where we disable validation, and migrate known config spec changes. As an alternative we could never change the config spec and only introduce new items. Longer term that could get quite messy though.

Thanks for the code snippet, I actually had a bit of a look for getting that info, but gave up on it.

feerrenrut commented:
@jcsteh I have updated this to include a config upgrade mechanism. There are a few other things I need to address on this branch but I would be interested to hear what you think of commit bc9e97d

feerrenrut commented:
@jcsteh I have pushed a couple of new changes to this.
- I have added a helper to be able to get validation parameters, this is helpful for setting the limits on the GUI code EG the min and max for spinControls. Now these values are not repeated.
- New / empty config files were previously having their schema version set to 0 and then being upgraded through to the latest. This upgrading a blank config through all steps is unnecessary since the latest schema should not require any previous, and has the potential to cause problems in the future if there are upgrade steps that rely on certain keys being present in a particular version which aren't.
- We talked about archiving the old profiles, rather than this I am putting the config into the log. Yes, this will increase the size of the log and is a bit verbose but it means that if there is a problem there is a chance that we can get the config back / get an example of what caused the problem.

jcsteh commented on user_docs/en/userGuide.t2t:
```diff
@@ -916,6 +916,9 @@ This option allows the word that is under the cursor to be displayed in non-cont
 It applies to the system caret and review cursor, but not to the selection indicator.
 
+==== Blink Cursor ====
+This option allows the braille cursor to blink. If blinking is turned off the braille cursor will just be static.
```
1. Please split this onto two lines, one per sentence.
2. I understand what you mean by "static", but I'm not sure many users will. Perhaps we should say "If blinking is turned off, the cursor will be constantly shown." Or we could change "constantly shown" to "always up" (i.e. the cursor dots are always up on the display).

jcsteh reviewed (changes requested):
1. I'm very probably missing something, but you mentioned at some point that you log the old and new configs. Where do you do this? I see the validation result get logged if it fails, but I'm guessing that doesn't log the actual values? Or does it? I'm confused on this one.
2. Do you think this is one of those situations where it might be worth rebasing and splitting this into smaller commits based on functionality? We'd then do a full merge rather than a squash. I'm wondering whether this might be clearer given that the config changes aren't strictly related to cursor blink, even though they're necessary for this change.

feerrenrut commented on source/config/profileUpgrader.py:
```diff
@@ -0,0 +1,67 @@
+
+def upgrade(profile, validator):
+	# when profile is none or empty we can still validate. It should at least have a version set.
+	log.info("Checking schema version of config:\n%s" % profile)
```
This is the one that prints out the profile / config in full before the upgrade process starts. This should stay as `info`.

feerrenrut replied:
In reply to your general comment about logging the config values, there is no where that logs the config after the upgrade. However the config file itself servers that purpose. If there is an error during the upgrade then having the starting point should be enough for us to reproduce it.

When there is an error, the items that dont validate are added to the log. Here is an example of the log file when this happens:

> INFO - __main__ (03:03:44):
> Starting NVDA
> INFO - core.main (03:03:45):
> Config dir: C:\work\nvda\source\userConfig
> INFO - config.ConfigManager._loadConfig (03:03:45):
> Loading config: .\userConfig\nvda.ini
> INFO - config.profileUpgrader.upgrade (03:03:45):
> Checking schema version of config:
> {u'schemaVersion': u'1', u'upgrade': {}, u'update': {}, u'general': {u'showWelcomeDialogAtStartup': u'False', u'language': u'en', u'saveConfigurationOnExit': u'True', u'askToExit': u'False', u'playStartAndExitSounds': u'True', u'loggingLevel': u'DEBUG'}, u'speech': {u'synth': u'espeak', u'espeak': {u'voice': u'en', u'variant': u'max', u'rate': u'30', u'pitch': u'40', u'inflection': u'75', u'volume': u'20'}}, u'braille': {u'cursorBlinkRate': u'-1', u'noBraille': {u'port': u''}}, u'keyboard': {u'useCapsLockAsNVDAModifierKey': u'True'}, u'documentFormatting': {u'autoLanguageSwitching': u'True'}, u'mouse': {u'enableMouseTracking': u'False', u'audioCoordinatesOnMouseMove': u'False', u'audioCoordinates_detectBrightness': u'False'}, u'speechViewer': {u'showSpeechViewerAtStartup': u'True', u'x': u'68', u'y': u'44', u'width': u'500', u'height': u'500', u'displays': [u'(1920, 1080)'], u'autoPositionWindow': u'False'}}
> ERROR - config.ConfigManager._initBaseConf (03:03:45):
> Error loading base configuration
> Traceback (most recent call last):
>   File "config\__init__.py", line 372, in _initBaseConf
>     profile = self._loadConfig(fn) # a blank file will be created if fn does not exist
>   File "config\__init__.py", line 400, in _loadConfig
>     profileUpgrader.upgrade(profile, self.validator)
>   File "config\profileUpgrader.py", line 24, in upgrade
>     _doValidation(deepcopy(profile), validator) # copy the profile, since validating mutates the object
>   File "config\profileUpgrader.py", line 57, in _doValidation
>     raise ValueError(errorString)
> ValueError: Unable to validate config file after upgrade: Key cursorBlinkRate : the value "-1" is too small.
> Full result: (value of false means the key was not present)
> [([u'braille'], u'cursorBlinkRate', VdtValueTooSmallError('the value "-1" is too small.',))]
> INFO - config.ConfigManager._loadConfig (03:03:45):
> Loading config: None
> INFO - config.profileUpgrader.upgrade (03:03:45):
> Checking schema version of config:
> {}
> INFO - core.main (03:03:45):
> NVDA version source-i6470-LimitCursorBlinkRate-076259c
> INFO - core.main (03:03:45):
> Using Windows version 6.2.9200 workstation
> INFO - core.main (03:03:45):
> Using Python version 2.7.11 (v2.7.11:6d1b6a68f775, Dec  5 2015, 20:32:19) [MSC v.1500 32 bit (Intel)]
> INFO - core.main (03:03:45):
> Using comtypes version 0.6.2
> INFO - synthDrivers.espeak.SynthDriver.__init__ (03:03:46):
> Using eSpeak version 1.48.15  16.Apr.15
> INFO - synthDriverHandler.setSynth (03:03:46):
> Loaded synthDriver espeak
> INFO - core.main (03:03:46):
> Using wx version 3.0.2.0 msw (classic)
> INFO - braille.initialize (03:03:46):
> Using liblouis version 3.0.0
> INFO - braille.BrailleHandler.setDisplayByName (03:03:46):
> Loaded braille display driver noBraille, current display has 0 cells.
> INFO - brailleInput.initialize (03:03:46):
> Braille input initialized
> WARNING - core.main (03:03:46):
> Java Access Bridge not available
> INFO - _UIAHandler.UIAHandler.MTAThreadFunc (03:03:46):
> UIAutomation: IUIAutomation3
> INFO - core.main (03:03:46):
> NVDA initialized
> INFO - config.ConfigManager.save (03:03:49):
> Base configuration saved
> INFO - core.main (03:03:55):
> Exiting
> INFO - config.ConfigManager.save (03:03:55):
> Base configuration saved
> INFO - __main__ (03:03:57):
> NVDA exit

jcsteh replied:
Sorry. My bad; didn't read it properly. I was confused because I saw "Checking schema version of config" and assumed it was just going to be a version number or something, not the config itself. If I'd looked more closely, I would have seen the profile variable afterwards. :)

I am a bit dubious about logging the entire config all the time. I totally understand why we'd log it before an upgrade, but unless I'm missing something, this will get logged on every load. It's true that having the config might be useful for debugging in some cases, but I feel it'd be appropriate for us to provide instructions to the user in that case; e.g. setting to level debug.

feerrenrut replied:
As discussed, there are two reasons to log the config on every load:
- Log the config before there is any chance of something going wrong while accessing the config (eg while getting the current schema version)
- Log the config so that we are able to diagnose mis-configurations of NVDA

The second of these goals is not really related to config upgrades. We will move that out of the upgrader and do it at debug level. The first goal, we will acheive by wrapping the first checks in a `try` and only logging at info level if there is an exception.

feerrenrut committed:
Introduce config/profile schema upgrade framework

When a schema is updated, config files (base config and profiles)
can be upgraded to meet the new schema. The `schemaVersion` (in
configSpec.py) should be incremented, and a new function should be added
to the `profileUpgradeSteps.py` file to make the changes required to
config files before they are loaded.

After all upgrade steps are made the file is validated, and any errors
will be in the log.

feerrenrut committed:
Fix #6470, limit the minimum blink rate to 200ms

Introduce a way to use the GUI to disable cursor blinking.

Rely on config update step to fix min values

The config schema updates should now fix values below the minimum.
Use validation data from config for min / max values in spin controls

jcsteh commented on source/config/__init__.py:
```diff
@@ -10,14 +16,18 @@
 from validate import Validator
-from logHandler import log
+from logHandler import log, levelNames
+from logging import DEBUG
```
This is also accessible as `log.DEBUG` (as opposed to the method `log.debug`). I thought I'd let you know in case you found it nicer to read, but I'm also happy for it to be left as is.

derekriemer commented on source/config/__init__.py:
```diff
@@ -914,6 +762,21 @@ def saveProfileTriggers(self):
+		spec = conf.spec
+		for nextKey in keyPath:
+			spec = spec[nextKey]
+		return conf.validator._parse_with_caching(spec)[2][validationParameter]
```
Is this a bug? (Hint: __many__ for voice settings).

derekriemer replied:
What i mean here is that in parts of the config, we have the parameter 
`[[__many__]]`
and I don't see a way to get the min or max from those. Could this be triggering the bug that was mentioned on NVDA-devel? [1] and [2]

[1]: http://nabble.nvda-project.org/Re-nvda-next-6157f43e-Incubates-6558-td41486.html
[2]: http://nabble.nvda-project.org/Latest-next-seems-to-crash-on-restart-locking-up-computer-tp41490.html

jcsteh replied:
I don't think this could be the cause of any current problems:

1. This is a new function that is only being used in one place right now (cursor blink rate) and that place doesn't involve a `__many__`.
2. Even though `__many__` is used for speech synths, synths build their own specific config spec which overrides this.

feerrenrut replied:
The issue shown in the links to the NVDA-devel list were caused by a merge error (my fault). A line was missed from the config spec.

In order to get the validation parameters from a `[[__many__]]` section the specific name of the section must be used. As you would if you were to get a config value from that section. For an example of this see [Line 511 in source/gui/settingsDialogs.py](https://github.com/nvaccess/nvda/pull/6558/files/820c02fb95784fb20215dc86e67f540d068ae602#diff-597828d33ba584f58afddcc26b80a992R511) which looks like: 
`min=int(config.conf.getConfigValidationParameter(["speech", getSynth().name, "capPitchChange"], "min"))`

derekriemer commented on user_docs/en/userGuide.t2t:
```diff
@@ -916,6 +916,10 @@ This option allows the word that is under the cursor to be displayed in non-cont
+
 ==== Cursor Blink Rate (ms) ====
 This option is a numerical field that allows you to change the blink rate of the cursor in milliseconds.
 
```
Should the minimum value be mentioned?
